### PR TITLE
AWS - Tag support for Elasticsearch

### PR DESCRIPTION
**Pull request by paultyng**
Since #3619 seems to have stalled, taking another stab at this.  I'm very new to Go so I'm sure this isn't that idiomatic, but I've tried to copy the patterns for tags from other resources.

The weird thing with Elasticsearch's API is that it doesn't allow passing tags in a create, its always a separate call.

I still need to write the tests for the tags on the resource (I have tests written for the tagging methods though).

### CODE REVIEW

**Comment by paultyng** [comment]
This now has testing for tags.

I need to run some additional runtime tests but its closer.

**Comment by paultyng** [comment]
Runtime testing on my template was successful, including just incremental tag update.  This is ready for review.

File: builtin/providers/aws/resource_aws_elasticsearch_domain.go [hunk 1]
```diff
@@ -37,13 +37,9 @@ func resourceAwsElasticSearchDomain() *schema.Resource {
 				ForceNew: true,
 				ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
 					value := v.(string)
-					if !regexp.MustCompile(`^[0-9A-Za-z]+`).MatchString(value) {
+					if !regexp.MustCompile(`^[a-z][0-9a-z\-]{2,27}$`).MatchString(value) {
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_elasticsearch_domain.go]: @paultyng why has this ValidateFunc changed?

**Comment by stack72** [comment]
@paultyng this looks good - as soon as I can understand why the ValidateFunc changed :)

The tests look good as well

```
make testacc TEST=./builtin/providers/aws TESTARGS='-run=AWSElasticSearch_tags'
==> Checking that code complies with gofmt requirements...
/Users/stacko/Code/go/bin/stringer
GO15VENDOREXPERIMENT=1 go generate $(GO15VENDOREXPERIMENT=1 go list ./... | grep -v /vendor/)
TF_ACC=1 GO15VENDOREXPERIMENT=1 go test ./builtin/providers/aws -v -run=AWSElasticSearch_tags -timeout 120m
=== RUN   TestAccAWSElasticSearch_tags
--- PASS: TestAccAWSElasticSearch_tags (1184.17s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/aws	1184.192s
```

File: builtin/providers/aws/resource_aws_elasticsearch_domain.go (hunk 1 above)
Comment (paultyng) [inline builtin/providers/aws/resource_aws_elasticsearch_domain.go]: Those are the actual rules for es domain names, I had it failing in the API call locally if I didn't meet those.  Let me dig up the docs, one sec.

Comment (paultyng) [inline builtin/providers/aws/resource_aws_elasticsearch_domain.go]: http://docs.aws.amazon.com/sdk-for-go/api/service/elasticsearchservice.html#type-CreateElasticsearchDomainInput

> Domain names must start with a letter or number and can contain the following characters: a-z (lowercase), 0-9, and - (hyphen).

The text from the actual console is:

> The name must start with a lowercase alphabet and be at least 3 and no more than 28 characters long. Valid characters are a-z (lowercase letters), 0-9, and - (hyphen).

And when I try to start with a number it doesn't work.

Feel free to cherry pick around this, or I can rip it out, I just hit this bug as well.

**Comment by paultyng** [comment]
Added inline responses above.  Basically bringing the client side validation in sync with the actual rules to fail faster on issues there.

**Comment by jen20** [comment]
:+1: for merging the fix for validation, but we should note this as a second CHANGELOG entry with the same PR number to avoid confusion.

**Comment by stack72** [comment]
Thanks @paultyng - makes sense!

@jen20 will add the extra entry to the changelog

Paul

**Pull request merged**